### PR TITLE
 [ncp] add support for clearing all/subset of counters

### PR DESCRIPTION
**Pull request by abtink**
This PR contains 3 related commits:

 **[ncp] add support for clearing all/subset of counters**
This commit adds support in NCP/spinel for clearing counters. It also adds new `PROP_CNTR_ALL_IP_COUNTERS` property to get all IPv6 counters. All (MAC, MLE, IPv6, and NCP) counters can be cleared by writing (`PROP_VALUE_SET` with any value) to `SPINEL_PROP_CNTR_RESET` property. A specific subset of counters (MAC, MLE, IP) can be cleared by writing (any value) to the corresponding spinel property (`PROP_CNTR_ALL_MAC_COUNTERS`, `PROP_CNTR_MLE_COUNTERS`, or    `PROP_CNTR_ALL_IP_COUNTERS`).

**[link] add public OT API otLinkResetCounters to reset MAC counters**

**[mesh-forwarder] add method ResetCounters to clear IP counter**
This commit also adds a public API `otThreadResetIp6Counters()`.

### CODE REVIEW

File: src/ncp/ncp_base_mtd.cpp [hunk 1]
```diff
@@ -2523,6 +2530,47 @@ template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_CNTR_MLE_COUNTERS>(vo
...
+
+    return OT_ERROR_NONE;
+}
+
+template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_CNTR_ALL_IP_COUNTERS>(void)
+{
+    otError             error    = OT_ERROR_NONE;
+    const otIpCounters *counters = otThreadGetIp6Counters(mInstance);
+
+    if (counters == NULL)
```
Comment (LuDuda) [inline src/ncp/ncp_base_mtd.cpp]: Shouldn't be assert instead?

Comment (abtink) [reply src/ncp/ncp_base_mtd.cpp]: Changed it in new pushed commits (also added the PR # in the commit titles).

File: src/core/thread/mesh_forwarder.cpp [hunk 2]
```diff
@@ -141,6 +138,11 @@ void MeshForwarder::Stop(void)
     return;
 }
 
+void MeshForwarder::ResetCounters(void)
```
Comment (jwhui) [inline src/core/thread/mesh_forwarder.cpp]: nit: Mle::ResetCounters()` is inlined. Should we make these consistent?

**Comment by abtink** [comment]
Related PR in wpantund: https://github.com/openthread/wpantund/pull/442

**Pull request merged**